lib/MongoDB/ChangeStream.pm

```diff
@@ -0,0 +1,201 @@
+#
+#  Copyright 2009-2018 MongoDB, Inc.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+
+use strict;
+use warnings;
+package MongoDB::ChangeStream;
+
+# ABSTRACT: A stream providing update information for collections.
+
+use version;
+our $VERSION = 'v1.999.0';
+
+use Moo;
+use Try::Tiny;
+use MongoDB::Cursor;
+use MongoDB::Op::_Aggregate;
+use MongoDB::Error;
+use Safe::Isa;
+use MongoDB::_Types qw(
+    MongoDBCollection
+    ArrayOfHashRef
+);
+use Types::Standard qw(
+    InstanceOf
+    HashRef
+    Str
+);
+
+use namespace::clean -except => 'meta';
+
+has _cursor => (
+    is => 'rw',
+    isa => InstanceOf['MongoDB::QueryResult'],
+    lazy => 1,
+    builder => '_build_cursor',
+    clearer => '_clear_cursor',
+);
+
+has collection => (
+    is => 'ro',
+    isa => MongoDBCollection,
+    required => 1,
+);
+
+has options => (
+    is => 'ro',
+    isa => HashRef,
+);
+
+has pipeline => (
+    is => 'ro',
+    isa => ArrayOfHashRef,
+    required => 1,
+);
+
+has full_document => (
+    is => 'ro',
+    isa => Str,
+    predicate => '_has_full_document',
+);
+
+has _resume_token => (
+    is => 'rw',
+    init_arg => 'resume_after',
+    predicate => '_has_resume_token',
+    lazy => 1,
+    builder => '_build_resume_token',
+);
+
+sub BUILD {
+    my ($self) = @_;
+
+    # starting point is construction time instead of first next call
+    $self->_cursor;
+}
+
+sub _build_cursor {
+    my ($self) = @_;
+
+    my $pipeline = $self->pipeline;
+
+    my @pipeline = @{ $self->pipeline || [] };
```

Comment:
I think the `|| []` isn't necessary anymore given `required` and the type constraint.